File: Assignment2_Code/part1_section2/javascript/typeScopeDemo.js
// JavaScript: Demonstrating Type Systems, Scoping, and Closures
// Program: A configurable discount calculator using closures

// --- Type System Demonstration ---
// JavaScript is dynamically typed with implicit type coercion.
let price = 100;         // number
let taxRate = 0.08;      // number
let label = "Widget";    // string

// Arithmetic works as expected with numbers
let total = price + (price * taxRate);
console.log(`Item: ${label}, Price with tax: ${total}, Type: ${typeof total}`);

// Type coercion: the + operator concatenates when one operand is a string
let coerced = price + " dollars";
console.log(`Coerced result: ${coerced}, Type: ${typeof coerced}`);

// Loose equality vs strict equality
console.log(`\n"5" == 5 (loose):  ${("5" == 5)}`);    // true (coercion)
console.log(`"5" === 5 (strict): ${("5" === 5)}`);     // false (no coercion)


// --- Closures Demonstration ---
// A closure captures variables from its enclosing lexical scope.
function makeDiscountCalculator(discountPercent) {
    // Returns a closure that remembers 'discountPercent'
    return function(originalPrice) {
        let savings = originalPrice * (discountPercent / 100);
        return originalPrice - savings;
    };
}

let tenPercentOff = makeDiscountCalculator(10);
let twentyFiveOff = makeDiscountCalculator(25);

console.log(`\n10% off $80: $${tenPercentOff(80).toFixed(2)}`);
console.log(`25% off $80: $${twentyFiveOff(80).toFixed(2)}`);


// --- Scoping Demonstration ---
// JavaScript has function scope (var) and block scope (let/const)
var globalVar = "I am global (var)";

function outerFunction() {
    let enclosingVar = "I am in the enclosing scope (let)";

    function innerFunction() {
        let localVar = "I am local (let)";
        console.log(`  Inner sees localVar: ${localVar}`);
        console.log(`  Inner sees enclosingVar: ${enclosingVar}`);
        console.log(`  Inner sees globalVar: ${globalVar}`);
    }

    innerFunction();
    // localVar is NOT accessible here (block-scoped with let)
}

console.log("\nScoping demonstration:");
outerFunction();


// --- Block Scope: var vs let ---
// 'var' is function-scoped (leaks out of blocks)
for (var j = 0; j < 3; j++) {
    var loopVarVar = j;
}
console.log(`\nAfter loop (var): j = ${j}, loopVarVar = ${loopVarVar}`);

// 'let' is block-scoped (stays inside the block)
for (let k = 0; k < 3; k++) {
    let loopVarLet = k;
}
// Uncommenting the next line would cause a ReferenceError:
// console.log(`After loop (let): k = ${k}`);
console.log("After loop (let): k and loopVarLet are NOT accessible (block-scoped)");

// --- var hoisting demonstration ---
console.log(`\nHoisted varDemo before assignment: ${varDemo}`); // undefined, not an error
var varDemo = "now assigned";
console.log(`varDemo after assignment: ${varDemo}`);

File: Assignment2_Code/part1_section2/python/type_scope_demo.py
# Python: Demonstrating Type Systems, Scoping, and Closures
# Program: A configurable discount calculator using closures

# --- Type System Demonstration ---
# Python is dynamically typed; variable types are determined at runtime.
price = 100          # int
tax_rate = 0.08      # float
label = "Widget"     # str

# Implicit type coercion: int + float produces float
total = price + (price * tax_rate)
print(f"Item: {label}, Price with tax: {total}, Type: {type(total)}")

# Dynamic typing allows reassignment to a different type
price = "fifty"
print(f"Reassigned price to string: {price}, Type: {type(price)}")
price = 100  # reset for later use


# --- Closures Demonstration ---
# A closure captures variables from its enclosing scope.
def make_discount_calculator(discount_percent):
    """Factory function that returns a closure."""
    def apply_discount(original_price):
        # 'discount_percent' is captured from the enclosing scope
        savings = original_price * (discount_percent / 100)
        return original_price - savings
    return apply_discount

ten_percent_off = make_discount_calculator(10)
twenty_five_off = make_discount_calculator(25)

print(f"\n10% off $80: ${ten_percent_off(80):.2f}")
print(f"25% off $80: ${twenty_five_off(80):.2f}")


# --- Scoping Demonstration ---
# Python uses LEGB rule: Local, Enclosing, Global, Built-in
global_var = "I am global"

def outer_function():
    enclosing_var = "I am in the enclosing scope"

    def inner_function():
        local_var = "I am local"
        print(f"  Inner sees local_var: {local_var}")
        print(f"  Inner sees enclosing_var: {enclosing_var}")
        print(f"  Inner sees global_var: {global_var}")

    inner_function()
    # Uncommenting the next line would cause a NameError:
    # print(local_var)

print("\nScoping demonstration:")
outer_function()


# --- Block Scope Behavior ---
# Python does NOT have block-level scoping for if/for/while
for i in range(3):
    loop_var = i

# 'loop_var' and 'i' are still accessible here
print(f"\nAfter loop, i = {i}, loop_var = {loop_var}")

if True:
    block_var = "created inside if"

print(f"After if block, block_var = {block_var}")
